srun jobID docs (#1945)
rcgsheffield/sheffield_hpc: hpc/scheduler/advanced_job_analysis_and _profiling.rst

.. _advanced_job_profiling_and_analysis:

Advanced Job Profiling and Analysis
===================================

.. tip::

    For majority of the users the :ref:`job_submission_control` page and the output log files you produce should suffice in analysing the perfomance of a batch job. This section is for users who want an even more granular view of the perfomance of their batch jobs by observing what is happening on the node their job is running. 

.. warning::

    Since your job has a fixed set of  memory and CPU resources, carrying out resource hungry operations might lead to the scheduler killing your batch job due to errors like ``out of memory``. 

    Abuse of this feature to carry out tasks that are not profiling and perfomance analysis of your running batch job might lead to your account being suspended.

> Accessing A Running Single Node Slurm Batch job

Reviewer: ```suggestion
Accessing A Running Single-Node Slurm Batch job
```

> Accessing A Running Single-Node Slurm Batch job

Reviewer 2: ```suggestion
Accessing a Running Single-Node Slurm Batch job
```

===============================================

> In some cases, you might want to access the resources of an existing scheduler allocation, for example a single node batch job that is in the running state. After allocating resources with ``sbatch`` you can access the resources (node) of this allocation with:

Reviewer: How about this instead?

```suggestion
In some cases, you might want to interact with a batch job in the ``RUNNING``  state (e.g. for fault-finding, debugging or profiling purposes).  You can start an interactive session within the resource allocation (memory and CPU cores on particular nodes) associated with the job with:
```

.. code-block:: shell

    srun --jobid=<JOBID> --pty /usr/bin/bash

> The command creates a new job step and starts interactive bash shell session within a previously allocated job, allowing you to interact with the resources allocated to that job.

Reviewer: ```suggestion
The command creates a new Job Step in the batch job with ID ``<JOBID>`` and starts interactive bash shell session within a that Job Step, allowing you to interact with the resources allocated to that job.
```

Reviewer: Also, do we want to link to reference info that explains what a Job Step is?

> The command creates a new Job Step in the batch job with ID ``<JOBID>`` and starts interactive bash shell session within a that Job Step, allowing you to interact with the resources allocated to that job.

Reviewer 2: ```suggestion
The command creates a new Job Step in the batch job with ID ``<JOBID>`` and starts an interactive bash shell session within that Job Step, allowing you to interact with the resources allocated to that job.
```

> If all the allocated CPU resources are already used, srun will prohibit the new job step the access to the resources. However, the argument ``--overlap`` can be passed to srun to allow job steps to overlap on the resources.

Reviewer: ```suggestion
If all the allocated CPU resources are already used, ``srun`` will prohibit the new Job Step the access to the resources. However, the argument ``--overlap`` can be passed to ``srun`` to allow Job Steps to overlap on the resources.
```

> If all the allocated CPU resources are already used, ``srun`` will prohibit the new Job Step the access to the resources. However, the argument ``--overlap`` can be passed to ``srun`` to allow Job Steps to overlap on the resources.

Reviewer 2: ```suggestion
If all the allocated CPU resources are already used, ``srun`` will prohibit the new Job Step to access those resources. However, the argument ``--overlap`` can be passed to ``srun`` to allow Job Steps to share those resources.
```

.. code-block:: shell

    srun --jobid=<JOBID> --overlap --pty /usr/bin/bash

Once you are in the interactive session you can see the process IDs associated with your job by typing:

.. code-block:: shell

    scontrol listpids |grep <JOBID>

    or

    ps -u $USER

> start profiling and analysing the perfomance of the node and the job by using commands such as:

Reviewer: ```suggestion
Start profiling and analysing the perfomance of the node and the job by using commands such as:
```

.. code-block:: shell

    ps

    nvidia-smi

    top

    lsof

> Accessing A Running Multi Node Slurm Batch job

Reviewer: ```suggestion
Accessing A Running Multi-Node Slurm Batch job
```

> Accessing A Running Multi-Node Slurm Batch job

Reviewer 2: ```suggestion
Accessing a Running Multi-Node Slurm Batch job
```

==============================================

> In the scenario you are running a multi node slurm job you can use ``squeue`` to see the nodes your job is using:

Reviewer: ```suggestion
In the scenario you are running a multi-node Slurm job you can use ``squeue`` to see the nodes your job is using:
```

.. code-block:: shell

>     squeue --user=$USER

Reviewer: ```suggestion
    squeue --me
```

Reviewer 3: Agree with this, I believe we also have quite a lot of the same change to make in the job submission pages.

Will look at this now.

**Example output:**

.. code-block:: shell

>     squeue --user=$USER

Reviewer: ```suggestion
    squeue --me
```
